File: android/app/src/main/kotlin/com/travelexpenses/android/init/DefaultCategoryInitializer.kt
package com.travelexpenses.android.init

import com.travelexpenses.event.ExpenseEvent
import com.travelexpenses.repository.CategoryRepository
import com.travelexpenses.repository.EventLogRepository
import com.travelexpenses.validation.DefaultCategories
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.datetime.Clock
import java.util.UUID

/**
 * Initializes default categories on first launch.
 * Checks if any categories exist, and if not, creates events for all 9 defaults.
 * Uses a mutex to prevent double-initialization from concurrent calls.
 */
class DefaultCategoryInitializer(
    private val categoryRepo: CategoryRepository,
    private val eventLogRepo: EventLogRepository,
    private val deviceId: String,
    private val scope: CoroutineScope,
) {
    private val mutex = Mutex()

    fun initializeIfNeeded() {
        scope.launch {
            mutex.withLock {
                val existing = categoryRepo.getAllCategories()
                if (existing.isNotEmpty()) return@withLock

                val now = Clock.System.now()
                for (category in DefaultCategories.all) {
                    val seqNum = eventLogRepo.count() + 1
                    val event = ExpenseEvent.CategoryCreated(
                        eventId = UUID.randomUUID().toString(),
                        timestamp = now,
                        sequenceNumber = seqNum,
                        deviceId = deviceId,
                        category = category,
                    )
                    eventLogRepo.append(event)
                }
            }
        }
    }
}

File: ios/TravelExpenses/DI/DefaultCategoryInitializer.swift
import Foundation
import Shared

/// Seeds the 9 default categories on first launch via CategoryCreated events.
/// Must be called once after the ServiceContainer is ready.
@MainActor
final class DefaultCategoryInitializer {

    private let categoryRepo: SqlDelightCategoryRepository
    private let eventLogRepo: SqlDelightEventLogRepository
    private let deviceId: String

    init(container: ServiceContainer) {
        self.categoryRepo = container.categoryRepo
        self.eventLogRepo = container.eventLogRepo
        self.deviceId = container.deviceId
    }

    func initializeIfNeeded() {
        Task {
            do {
                let existing = try await categoryRepo.getAllCategories()
                guard existing.isEmpty else { return }

                let now = Kotlinx_datetimeClock.companion.System.now()
                let defaults = DefaultCategories.shared.all

                for (index, category) in defaults.enumerated() {
                    let seqNum = try await eventLogRepo.count() + Int64(index + 1)
                    let event = ExpenseEvent.CategoryCreated(
                        eventId: UUID().uuidString,
                        timestamp: now,
                        sequenceNumber: seqNum,
                        deviceId: deviceId,
                        category: category
                    )
                    try await eventLogRepo.append(event: event)
                }
            } catch {
                // Non-fatal – categories will be absent but app still functions
                print("[DefaultCategoryInitializer] Failed: \(error)")
            }
        }
    }
}
